tracing/middlewares.go

```diff
@@ -9,26 +9,32 @@ import (
 // InjectHTTPServerSpan returns a go-kit endpoint.Middleware that injects a server
 // span into the `next` context.
 //
-// Note, this depends on the edge context headers already being set on the
-// context object.  This can be done by adding httpbp.PopulateRequestContext as
-// a ServerBefore option when setting up the request handler for an endpoint.
+// Starts the server span before calling the `next` endpoint and stops the span
+// after the endpoint finishes.
+// If the endpoint returns an error, that will be passed to span.Stop. If the
+// response implements ErrorResponse, the error returned by Err() will not be
+// passed to span.Stop.
+//
+// Note, this function depends on the edge context headers already being set on
+// the context object.  This can be done by adding httpbp.PopulateRequestContext
+// as a ServerBefore option when setting up the request handler for an endpoint.
 func InjectHTTPServerSpan(name string) endpoint.Middleware {
-	return func(next endpoint.Endpoint) endpoint.Endpoint {
-		return func(ctx context.Context, request interface{}) (interface{}, error) {
-			ctx, _ = StartSpanFromHTTPContext(ctx, name)
-			return next(ctx, request)
-		}
-	}
+	return InjectHTTPServerSpanWithTracer(name, nil)
 }
 
 // InjectHTTPServerSpanWithTracer is the same as InjectHTTPServerSpan except it
 // uses StartSpanFromHTTPContextWithTracer to initialize the server span rather
 // than StartSpanFromHTTPContext.
-func InjectHTTPServerSpanWithTracer(tracer *Tracer, name string) endpoint.Middleware {
+func InjectHTTPServerSpanWithTracer(name string, tracer *Tracer) endpoint.Middleware {
 	return func(next endpoint.Endpoint) endpoint.Endpoint {
-		return func(ctx context.Context, request interface{}) (interface{}, error) {
-			ctx, _ = StartSpanFromHTTPContextWithTracer(ctx, name, tracer)
-			return next(ctx, request)
+		return func(ctx context.Context, request interface{}) (response interface{}, err error) {
+			ctx, span := StartSpanFromHTTPContextWithTracer(ctx, name, tracer)
+			defer func() {
+				span.Stop(ctx, err)
```

Comment:
The "simplified" way didn't work because `ctx` and `err` were being evaluated when `defer` executed rather than when the `defer`-ed `span.Stop` is executed.